design/backup-performance-improvements.md

```diff
@@ -0,0 +1,211 @@
+`SetupWithManager` will also add the input channel to the `itemBackupper` so that it will be available during backup processing.
+The func `RunItemBlockWorkers` will create the `ItemBlockWorkerPool` with a shared buffered input channel (fixed buffer size) and start `workers` gororoutines which will each call `processItemBlocksWorker`.
+The `processItemBlocksWorker` func (run by the worker goroutines) will read from `itemBlockChannel`, call `BackupItemBlock` on the retrieved `ItemBlock`, and then send the return value to the retrieved `returnChan`, and then process the next block.
+
+#### Modify ItemBlock processing loop to send ItemBlocks to the worker pool rather than backing them up directly
+
+The ItemBlock processing loop implemented in Phase 1 will be modified to send each newly-created ItemBlock to the shared channel rather than calling `BackupItemBlock` inline, using a WaitGroup to manage in-process items. A separate goroutine will be created to process returns for this backup. After completion of the ItemBlock processing loop, velero will use the WaitGroup to wait for all ItemBlock processing to complete before moving forward.
+
+#### Synchronize access to the BackedUpItems map
+
+Velero uses a map of BackedUpItems to track which items have already been backed up. This prevents velero from attempting to back up an item more than once, as well as guarding against creating infinite loops due to circular dependencies in the additional items returns. Since velero will now be accessing this map from the parallel goroutines, access to the map must be synchronized with mutexes.
+
+#### V3 vs V1/2 BackupItemAction plugins registered
+
+Full item block functionality is only possible if all registered BIA plugins implement the v3 interface. In phase 1 work, if any v1 or v2 plugins are registered, then v3 plugins are adapted to v2 and we treat every item returned from the collector as an ItemBlock of one. Because we are not able to track inter-item dependencies in this operation mode, we will also need to ensure that only one worker is active in processing these ItemBlocks of size 1 -- since we're not tracking dependencies, we must process items in the order returned by the collector to ensure that PVCs are backed up with their Pods, etc. Therefore, the worker count will be 1, regardless of configuration, in this scenario. A warning should be logged when starting the worker pool when this happens.
+
+## Alternatives considered
+
+### Per-backup worker pool
+
+The current design makes use of a permanent worker pool, started at backup controller startup time. With this design, when we follow on with running multiple backups in parallel, the same set of workers will take ItemBlock inputs from more than one backup. Another approach that was initially considered was a temporary worker pool, created while processing a backup, and deleted upon backup completion. 
```

Comment:
what are your thoughts on API server throttling due to this parallelism?

Reply:
From an APIServer point of view, I think permanent vs. temporary worker pools should be basically equivalent. Also, from an APIServer point of view, Restore is more likely to hit throttling than backup, since every item restore requires at least two APIServer calls. On the backup side, the Item Collector uses list calls to grab the GVRs included in the backup, so individual calls are only needed for items returned as AdditionalItems that weren't in the original list or by BIA plugins that make their own API calls.

All of this is to say that I suspect even with parallel processing, most backups will make less frequent APIServer calls than an equivalent restore. It doesn't guarantee that we won't run into throttling issues, but in cases where we do, we're probably already hitting them on restore. As for what we need to do, I'm not 100% sure. But since that's as much a potential issue with current restore code, would it make sense to address that for both Backup and Restore as a separate effort?

Reply:
It makes sense to tackle it separately for backup and restore.
But I think it would be good to track an effort to understand the current pressure velero puts on the API server.
For example, can several List calls lead to latency in API server. 
Basically some benchmark to understand current behaviour. Which can be compared to this work later one.

<Sidenote>
One other worry I have is that velero will become a spikey workload. 
Whenever backup starts, at once all workers might spike in mem/ cpu. If velero is running on a constrained node, this might lead to pod restarts. We have seen this for bunch of our customers.

If our eventual goal is to support parallel backups, should we consider spawning a new pod/ job for each backup job which gets scheduled on appropriate node and runtime and is less likely to hit issues.  

Reply:
From an APIServer point of view, parallel backup shouldn't change much -- we do all of the list calls at the beginning via the item collector before handing anything off to workers. The only APIServer calls that will result from the parallel workers will be plugin-related. While it's possible that these will happen more quickly with the new design, they will still be significantly less frequent than what we see on Restore, where every single item restored has 2 APIServer calls -- one Create, and one Patch.

Reply:
If we want to consider spawning pods/jobs for parallel backup actions, that is something to consider when we design the "run 2 backups at once" feature, which is a non-goal for this design. I'm not sure the added complexity is warranted there, though -- but that's a conversation to have when we're considering that design.

Reply:
alright
seems fair for now.